### PR TITLE
Fowrawrding Refs

### PR DESCRIPTION
**Pull request by hosseinAMD**

### CODE REVIEW

**Comment by sorousht** [comment]
I guess you need to get the latest changes of branch `master`, and merge it into your own branch which is `hosseinAMD-forwarding-refs-hosseinamd` by following the below steps:

```bash
git checkout master
git pull origin master
git checkout hosseinAMD-forwarding-refs-hosseinamd
git merge master
```

If you still have some troubles, let me know :)

**Comment by hosseinAMD** [comment]
> 
> 
> I guess you need to get the latest changes of branch `master`, and merge it into your own branch which is `hosseinAMD-forwarding-refs-hosseinamd` by following the below steps:
> 
> ```shell
> git checkout master
> git pull origin master
> git checkout hosseinAMD-forwarding-refs-hosseinamd
> git merge master
> git push [your-own-remote-name] hosseinAMD-forwarding-refs-hosseinamd
> ```
> 
> If you still have some troubles, let me know :)

@sorousht

I've ran every command you told me. but as you can see tests still fail
I can send you my translation .MD file to test if you have linux os!

**Comment by hosseinAMD** [comment]
@sorousht 

An if you check the logs you can find this line the reason of failing tests

`Cannot find module '@babel/compat-data/corejs3-shipped-proposals'`

It's the reason of failing all tests!

**Comment by hosseinAMD** [comment]
OK it solved with fetching upstream repository Finally! :)
Thanks @sorousht for your efforts

File: content/docs/forwarding-refs.md [hunk 1]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 
-Conditionally applying `React.forwardRef` when it exists is also not recommended for the same reasons: it changes how your library behaves and can break your users' apps when they upgrade React itself.
+استفاده شرطی از `React.forwardRef` در زمان وجود آن نیز به دلیل یکسان پیشنهاد نمی شود: چرا که چگونگی رفتار کتابخانه شما را تغییر می دهد و برنامه کاربرانتان را زمانی که ری‌اکت خود را ارتقا می دهد خراب می کند.
 
-## Forwarding refs in higher-order components {#forwarding-refs-in-higher-order-components}
+## Forwarding refs در کامپوننت های مرتبه بالاتر {#forwarding-refs-in-higher-order-components}
 
-This technique can also be particularly useful with [higher-order components](/docs/higher-order-components.html) (also known as HOCs). Let's start with an example HOC that logs component props to the console:
+این تکنیک در [کامپوننت های مرتبه بالاتر](/docs/higher-order-components.html) (شناخته شده به نام HOC ها) نیز، می تواند مفید باشد. بیایید با یک مثال از HOC که prop های کامپوننت را در کنسول، نمایش می دهد شروع کنیم:
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:50]:
```suggestion
این تکنیک در [کامپوننت های مرتبه بالاتر](/docs/higher-order-components.html) (شناخته شده به نام HOC ها) نیز، می تواند مفید باشد. بیایید با یک مثال از HOC که prop های کامپوننت را در کنسول، نمایش می‌دهد شروع کنیم:
```

File: content/docs/forwarding-refs.md [hunk 2]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 `embed:forwarding-refs/log-props-before.js`
 
-The "logProps" HOC passes all `props` through to the component it wraps, so the rendered output will be the same. For example, we can use this HOC to log all props that get passed to our "fancy button" component:
+
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:53]:
```suggestion
```

File: content/docs/forwarding-refs.md [hunk 3]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
+کامپوننت مرتبه بالاتر "logProps" کل `props` را به کامپوننتی که پوشش داده است منتقل می کند، بنابراین خروجی رندر شده یکسان خواهد بود. برای مثال، می توانیم از این HOC برای مشاهده تمامی prop هایی که به کامپوننت "fancy button" ما منتقل شده است استفاده کنیم:
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:54]:
```suggestion
کامپوننت مرتبه بالاتر "logProps" کل `props` را به کامپوننتی که پوشش داده‌است منتقل می‌کند، بنابراین خروجی رندر شده یکسان خواهد بود. برای مثال، می‌توانیم از این HOC برای مشاهده تمامی prop هایی که به کامپوننت "fancy button" ما منتقل شده‌است استفاده کنیم:
```

File: content/docs/forwarding-refs.md [hunk 4]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 `embed:forwarding-refs/fancy-button.js`
 
-There is one caveat to the above example: refs will not get passed through. That's because `ref` is not a prop. Like `key`, it's handled differently by React. If you add a ref to a HOC, the ref will refer to the outermost container component, not the wrapped component.
 
-This means that refs intended for our `FancyButton` component will actually be attached to the `LogProps` component:
+فقط یک اخطار در مورد مثال بالا وجود دارد: ref ها منتقل نمی شوند. به علت اینکه `ref` یک prop نیست. مانند `key`، توسط ری‌اکت به شکل متفاوتی کنترل می شود. اگر شما یک ref به HOC اضافه کنید، ref به جای کامپوننت پوشیده شده، به بیرونی ترین کامپوننت پوشاننده اشاره می کند.
+
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:59]:
```suggestion
```

File: content/docs/forwarding-refs.md [hunk 5]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
+
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:60]:
```suggestion
```

File: content/docs/forwarding-refs.md [hunk 6]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
+
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:61]:
```suggestion
```

File: content/docs/forwarding-refs.md [hunk 7]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 
-The "logProps" HOC passes all `props` through to the component it wraps, so the rendered output will be the same. For example, we can use this HOC to log all props that get passed to our "fancy button" component:
+
+کامپوننت مرتبه بالاتر "logProps" کل `props` را به کامپوننتی که پوشش داده است منتقل می کند، بنابراین خروجی رندر شده یکسان خواهد بود. برای مثال، می توانیم از این HOC برای مشاهده تمامی prop هایی که به کامپوننت "fancy button" ما منتقل شده است استفاده کنیم:
 `embed:forwarding-refs/fancy-button.js`
 
-There is one caveat to the above example: refs will not get passed through. That's because `ref` is not a prop. Like `key`, it's handled differently by React. If you add a ref to a HOC, the ref will refer to the outermost container component, not the wrapped component.
 
-This means that refs intended for our `FancyButton` component will actually be attached to the `LogProps` component:
+فقط یک اخطار در مورد مثال بالا وجود دارد: ref ها منتقل نمی شوند. به علت اینکه `ref` یک prop نیست. مانند `key`، توسط ری‌اکت به شکل متفاوتی کنترل می شود. اگر شما یک ref به HOC اضافه کنید، ref به جای کامپوننت پوشیده شده، به بیرونی ترین کامپوننت پوشاننده اشاره می کند.
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:58]:
```suggestion
فقط یک اخطار در مورد مثال بالا وجود دارد: ref ها منتقل نمی‌شوند. به علت اینکه `ref` یک prop نیست. مانند `key`، توسط ری‌اکت به شکل متفاوتی کنترل می‌شود. اگر شما یک ref به HOC اضافه کنید، ref به جای کامپوننت دربر گیرنده، به بیرونی ترین کامپوننت پوشاننده (container) اشاره می کند.
```

File: content/docs/forwarding-refs.md [hunk 8]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
+
+
+
+این به این معناست که ref تعیین شده برای `FancyButton`، در اصل به کامپوننت `LogProps` متصل خواهد شد:
 `embed:forwarding-refs/fancy-button-ref.js`
 
-Fortunately, we can explicitly forward refs to the inner `FancyButton` component using the `React.forwardRef` API. `React.forwardRef` accepts a render function that receives `props` and `ref` parameters and returns a React node. For example:
+
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:65]:
```suggestion
```

File: content/docs/forwarding-refs.md [hunk 9]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
+خوشبختانه، با استفاده از API های `React.forwardRef`، می توانیم به طور صریح ref ها را به کامپوننت داخلی `FancyButton` فوروارد کنیم. `React.forwardRef` یک تابع رندر که پارامترهای  `props` و `ref` را دریافت می کند و یک نود ری‌اکت را به عنوان خروجی می دهد را می پذیرد. برای مثال:
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:66]:
```suggestion
خوشبختانه، با استفاده از API های `React.forwardRef`، می توانیم به طور صریح ref ها را به کامپوننت داخلی `FancyButton` فوروارد کنیم. `React.forwardRef` یک تابع رندر که پارامترهای  `props` و `ref` را دریافت می‌کند و یک نود ری‌اکت را به عنوان خروجی می‌دهد را می‌پذیرد. برای مثال:
```

File: content/docs/forwarding-refs.md [hunk 10]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 `embed:forwarding-refs/log-props-after.js`
 
-## Displaying a custom name in DevTools {#displaying-a-custom-name-in-devtools}
+## نمایش یک نام سفارشی در DevTools {#displaying-a-custom-name-in-devtools}
 
-`React.forwardRef` accepts a render function. React DevTools uses this function to determine what to display for the ref forwarding component.
+`React.forwardRef` یک تابع رندر می پذیرد. React DevTools از این تابع برای بررسی اینکه چه چیزی را برای کامپوننت ref forwarding نمایش دهد استفاده می کند.
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:71]:
```suggestion
`React.forwardRef` یک تابع رندر می‌پذیرد. React DevTools از این تابع برای بررسی اینکه چه چیزی را برای کامپوننت ref forwarding نمایش دهد استفاده می‌کند.
```

File: content/docs/forwarding-refs.md [hunk 11]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 
-For example, the following component will appear as "*ForwardRef*" in the DevTools:
+برای مثال، کامپوننت زیر به عنوان "*ForwardRef*" در DevTools ظاهر می شود:
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:73]:
```suggestion
برای مثال، کامپوننت زیر به عنوان "*ForwardRef*" در DevTools ظاهر می‌شود:
```

File: content/docs/forwarding-refs.md [hunk 12]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 
 `embed:forwarding-refs/wrapped-component.js`
 
-If you name the render function, DevTools will also include its name (e.g. "*ForwardRef(myFunction)*"):
+اگر شما تابع رندر را نامگذاری کنید، DevTools نام آنرا نیز بکار می برد. (برای مثال  "*ForwardRef(myFunction)*"):
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:77]:
```suggestion
اگر شما تابع رندر را نامگذاری کنید، DevTools نام آن‌را نیز به‌کار می‌برد. (برای مثال  "*ForwardRef(myFunction)*"):
```

File: content/docs/forwarding-refs.md [hunk 13]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
...
 
 `embed:forwarding-refs/wrapped-component-with-function-name.js`
 
-You can even set the function's `displayName` property to include the component you're wrapping:
+شما حتی می توانید ویژگی `displayName` تابع را تنظیم کنید تا کامپوننتی که پوشش می دهید را شامل شود:
```
Comment (sorousht) [inline content/docs/forwarding-refs.md:81]:
```suggestion
شما حتی می‌توانید ویژگی `displayName` تابع را تنظیم کنید تا کامپوننتی که پوشش می‌دهید را شامل شود:
```

File: content/docs/forwarding-refs.md [hunk 14]
```diff
@@ -4,73 +4,80 @@ title: Forwarding Refs
 permalink: docs/forwarding-refs.html
```
[the remark this answers is not in the document]
Comment (hosseinAMD) [reply content/docs/forwarding-refs.md:4]: Resolved at 
[ecf44d0 ](https://github.com/reactjs/fa.reactjs.org/pull/150/commits/ecf44d0529751592e60df6841a290b70679230e9)

**Pull request merged**